### PR TITLE
support llava video

### PR DESCRIPTION
**Pull request by ZhangYuanhan-AI**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -327,6 +327,101 @@ response = client.chat.completions.create(
 print(response)
 ```
 
+```python
```
Comment (merrymercy) [inline README.md]: Revert the change of READM.md.

File: examples/quick_start/srt_example_llava_v.sh [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+#      ```
+#    - The K node:
+#      ```sh
+#      bash examples/quick_start/srt_example_llava_v.sh K K-1 YOUR_VIDEO_PATH YOUR_MODEL_PATH FRAMES_PER_VIDEO
+#      ```
+
+
+# Replace `K`, `YOUR_VIDEO_PATH`, `YOUR_MODEL_PATH`, and `FRAMES_PER_VIDEO` with your specific details.
+
+cd /mnt/bn/vl-research/workspace/yhzhang/sglang_video
```
Comment (merrymercy) [inline examples/quick_start/srt_example_llava_v.sh]: Can you move all your examples under `examples/usage/llava_video` per our offline discussion?

You can think of llava video as an advanced usage so it does not belong to `quick_start`

File: python/pyproject.toml [hunk 3]
```diff
@@ -20,7 +20,7 @@ dependencies = [
 
 [project.optional-dependencies]
 srt = ["aiohttp", "fastapi", "psutil", "rpyc", "torch", "uvloop", "uvicorn",
-       "zmq", "vllm>=0.4.2", "interegular", "pydantic", "pillow", "outlines>=0.0.27", "flashinfer>=0.0.4", "packaging"]
+       "zmq", "vllm>=0.4.2", "interegular", "pydantic", "pillow", "outlines>=0.0.27","packaging","opencv-python-headless","huggingface_hub","hf_transfer","outlines==0.0.34"]
```
Comment (merrymercy) [inline python/pyproject.toml]: there are two `outlines`?

File: python/sglang/api.py [hunk 4]
```diff
@@ -149,6 +150,10 @@ def image(expr: SglExpr):
     return SglImage(expr)
 
 
+def video(expr_1: SglExpr, expr_2: SglExpr):
```
Comment (merrymercy) [inline python/sglang/api.py]:
```suggestion
def video(path: str, num_frames: int):
```

File: python/sglang/backend/openai.py [hunk 5]
```diff
@@ -11,7 +11,6 @@
 
 try:
     import openai
-    import tiktoken
```
Comment (merrymercy) [inline python/sglang/backend/openai.py]: tiktoken is used here. Please do not delete it.

File: python/sglang/launch_server_llavavid.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+from sglang.srt.server import ServerArgs, launch_server
+
+import multiprocessing as mp
+
+import argparse
+
+# model_overide_args["image_token_index"] = 64002
+
+
+
```
Comment (merrymercy) [inline python/sglang/launch_server_llavavid.py]: Remove redundant empty lines

File: python/sglang/srt/utils.py [hunk 7]
```diff
@@ -19,6 +19,8 @@
 
 show_time_cost = False
 time_infos = {}
+import torch.distributed as dist
```
Comment (merrymercy) [inline python/sglang/srt/utils.py]: remove this?

File: python/sglang/srt/utils.py [hunk 8]
```diff
@@ -19,6 +19,8 @@
...
+from PIL import Image
```
Comment (merrymercy) [inline python/sglang/srt/utils.py]: move this to `decode_video_base64` so we can make `PIL` an optional dependency.
See `load_image` as an example

File: python/sglang/utils.py [hunk 9]
```diff
@@ -2,12 +2,19 @@
...
 import threading
 import urllib.request
+from concurrent.futures import ThreadPoolExecutor
 from io import BytesIO
 from json import dumps
 
+import cv2
+import numpy as np
 import requests
+from PIL import Image
```
Comment (merrymercy) [inline python/sglang/utils.py]: move 
`import cv2` and `from PIL import Image` under the function `encode_frame` so we can make them optional depenendcy

File: test/srt/test_jump_forward.py [hunk 10]
```diff
@@ -1,6 +1,7 @@
 import argparse
 from enum import Enum
 
+import sglang as sgl
```
Comment (merrymercy) [inline test/srt/test_jump_forward.py]: revert the change

**Review by merrymercy — CHANGES_REQUESTED** [review]

**Comment by ZhangYuanhan-AI** [comment]
Looks like there are no conflict at README.md

File: python/sglang/__init__.py [hunk 11]
```diff
@@ -46,7 +47,7 @@
     "gen_int",
     "gen_string",
     "image",
-    "select",
+    "video" "select",
```
Comment (merrymercy) [inline python/sglang/__init__.py]:
```suggestion
    "video",
    "select",
```

File: python/sglang/api.py [hunk 12]
```diff
@@ -151,6 +152,10 @@ def image(expr: SglExpr):
     return SglImage(expr)
 
 
+def video(path: str, num_frames: int):
+    return SglVideo(expr_1, expr_2)
```
Comment (merrymercy) [inline python/sglang/api.py]:
```suggestion
def video(path: str, num_frames: int):
    return SglVideo(str, num_frames)
```

File: python/sglang/utils.py [hunk 13]
```diff
@@ -2,13 +2,18 @@
...
 import threading
 import urllib.request
+from concurrent.futures import ThreadPoolExecutor
 from io import BytesIO
 from json import dumps
 
+import numpy as np
 import requests
 
+# from decord import VideoReader, cpu
```
Comment (merrymercy) [inline python/sglang/utils.py]:
```suggestion
```

**Review by merrymercy — COMMENTED** [review]
We are very close now, but I found several obvious errors that make the code unable to run. Could you fix them and do one test?

File: python/pyproject.toml [hunk 14]
```diff
@@ -20,7 +20,7 @@ dependencies = [
 
 [project.optional-dependencies]
 srt = ["aiohttp", "fastapi", "psutil", "rpyc", "torch", "uvloop", "uvicorn",
-       "zmq", "vllm>=0.4.2", "interegular", "pydantic", "pillow", "outlines>=0.0.27", "packaging"]
+       "zmq", "vllm>=0.4.2", "interegular", "pydantic", "pillow", "packaging", "huggingface_hub", "hf_transfer", "outlines>=0.0.34"，"opencv-python-headless"]
```
Comment (merrymercy) [inline python/pyproject.toml]:
```suggestion
       "zmq", "vllm>=0.4.2", "interegular", "pydantic", "pillow", "packaging", "huggingface_hub", "hf_transfer", "outlines>=0.0.34"]
```

Let us make opencv-python-headless an optional dependency because most text only models do not need it

**Comment by merrymercy** [comment]
Your PR makes llava-v1.6 not runnable. Could you do more tests and let me know it is fully ready for merge?
```
cd sglang/examples/quick_start
python3 srt_example_llava.py
```

Error:
image

**Comment by merrymercy** [comment]
@ZhangYuanhan-AI Thanks! It is merged
